### PR TITLE
[Fiber] Use synchronous scheduling by default

### PR DESCRIPTION
**Pull request by acdlite**
~~Work in progress~~ Ready for review
- [x] Implement synchronous work
- [x] Add host config to opt into synchronous scheduling
- [x] Batch nested updates
- [x] Turn on sync scheduling in ReactDOMFiber
- [x] Figure out which tests are failing as a result, and why
- [x] Expose `unstable_batchedUpdates` API

### CODE REVIEW

File: src/renderers/shared/fiber/ReactFiberScheduler.js [hunk 1]
```diff
@@ -404,22 +404,57 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
...
 
-  function scheduleUpdate(fiber: Fiber, priorityLevel : PriorityLevel): void {
+  function scheduleUpdate(fiber: Fiber, priorityLevel : ?PriorityLevel): void {
+    // Use priority context if no priority is provided
+    if (priorityLevel == null) {
+      priorityLevel = priorityContext !== null ?
+        priorityContext :
+        defaultPriorityContext;
+    }
+
```
Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: Probably better to perform the work here, rather than go through the whole scheduling dance.

**Comment by acdlite** [comment]
Just to be clear, I'm still working on this. Don't quite know what I'm doing yet :D

File: src/renderers/noop/ReactNoop.js [hunk 2]
```diff
@@ -210,6 +211,10 @@ var ReactNoop = {
     NoopRenderer.performWithPriority(AnimationPriority, fn);
   },
 
+  setDefaultPriority(priorityLevel: PriorityLevel) {
+    NoopRenderer.setDefaultPriority(priorityLevel);
+  },
```
Comment (sebmarkbage) [inline src/renderers/noop/ReactNoop.js]: Can we just make this low pri by default in ReactNoop and synchronous in ReactDOMFiber? So we don't have to expose this as an API?

Comment (acdlite) [inline src/renderers/noop/ReactNoop.js]: Oh duh

**Comment by acdlite** [comment]
Rebased on top of Dan's error boundary work

**Comment by gaearon** [comment]
It would be great if you could add a test in `ErrorBoundaries-test` for `setState` inside `batchedUpdates`, like this:

```js
batchedUpdates(() => {
  component.setState({ ... });
  throw something;
});
```

It needs to still perform the work despite throwing.

Not related to error boundaries per se but we have all tests about errors there now so makes sense to add it there.

**Comment by gaearon** [comment]
>Work in progress

It will be current once you commit!
(I just can't resist terrible Fiber puns.)

File: src/renderers/shared/fiber/ReactFiberClassComponent.js [hunk 3]
```diff
@@ -47,19 +46,19 @@ module.exports = function(scheduleUpdate : (fiber: Fiber, priorityLevel : Priori
       const updateQueue = fiber.updateQueue ?
         addToQueue(fiber.updateQueue, partialState) :
         createUpdateQueue(partialState);
-      scheduleUpdateQueue(fiber, updateQueue, LowPriority);
+      scheduleUpdateQueue(fiber, updateQueue);
```
Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberClassComponent.js]: I think I found the problem... If `scheduleUpdateQueue` is sync (not batched) then the update is flushed before `enqueueCallback` is ever called. Then `enqueueCallback` will add a callback to the queue but it won't ever get called until the next time the fiber is committed.

File: src/renderers/shared/fiber/ReactFiberScheduler.js [hunk 4]
```diff
@@ -62,9 +62,17 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
 
   const scheduleAnimationCallback = config.scheduleAnimationCallback;
   const scheduleDeferredCallback = config.scheduleDeferredCallback;
+  const useSyncScheduling = config.useSyncScheduling;
 
-  // The default priority to use for updates.
-  let defaultPriority : PriorityLevel = LowPriority;
+  // The priority level to use when scheduling an update.
+  let priorityContext : (PriorityLevel | null) = null;
```
Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: AFAIK having a variable switch between null and a number prevents some optimizations in V8. Can we use something like `-1` for this?

Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: You could stil keep the distinction in type system if you don't make it a part of the `PriorityContext` union and explicitly say `PriorityContext | NoPriorityContext`. Or something like it.

File: src/renderers/shared/fiber/ReactFiberScheduler.js [hunk 5]
```diff
@@ -498,6 +479,75 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
...
+
+      if (!shouldBatchUpdates) {
+        performSynchronousWork();
+      }
+    }
+  }
+
+  function performAndHandleErrors<A>(fn: (a: A) => void, a: A) {
+    try {
+      fn(a);
```
Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: Do we try to avoid first class functions unless we're dealing with browser or user code? I'm not sure, just curious what @sebmarkbage wants.

File: src/renderers/shared/stack/reconciler/__tests__/ReactCompositeComponentState-test.js [hunk 6]
```diff
@@ -148,72 +151,146 @@ describe('ReactCompositeComponent-state', () => {
...
+        ['componentDidUpdate-currentState', 'blue'],
+        ['componentDidUpdate-prevState', 'blue'],
+        ['forceUpdate', 'blue'],
+        // unmountComponent()
+        // state is available within `componentWillUnmount()`
+        ['componentWillUnmount', 'blue'],
+      ];
+    } else {
+      // There's a bug in the stack reconciler where setState callbacks inside
+      // componentWillMount aren't flushed properly
```
Comment (gaearon) [inline src/renderers/shared/stack/reconciler/__tests__/ReactCompositeComponentState-test.js]: Oh nice, we are fixing this?

Comment (acdlite) [inline src/renderers/shared/stack/reconciler/__tests__/ReactCompositeComponentState-test.js]: Yes we should fix this in the stack reconciler. I'll open an issue for it once I understand it better. (Not very familiar with how stack works.) @sebmarkbage has more details.

Comment (acdlite) [inline src/renderers/shared/stack/reconciler/__tests__/ReactCompositeComponentState-test.js]: The extent of my knowledge is: under certain circumstances in `componentWillMount`, enqueued callbacks aren't flushed, so they stick around until some other update causes them to flush.

File: src/renderers/shared/fiber/ReactFiberScheduler.js (hunk 4 above)
Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: Good call. I think I'll just get rid of `defaultPriorityContext` entirely. I added it when I thought we'd need a way to change the default on demand, but now that it's determined by the host config it's not necessary.

**Comment by acdlite** [comment]
Ready for review!

**Comment by acdlite** [comment]
@gaearon I'll look at the error boundary stuff once we get https://github.com/facebook/react/pull/8153 merged.

File: src/renderers/shared/fiber/ReactFiberScheduler.js (hunk 5 above)
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: Ideally we should avoid first-class functions because I'd like everything to be able to take advantage of inline expansion - and be easily portable to C. It also helps newcomers to follow code that you don't have to jump back and forth. Not sure if we'll even use first-class functions for scheduling in React Native because we're not tied to rAF/rIC. Might just move that part to the DOM renderer.

It'd probably be better to go in the opposite direction and unify deferred/animation work and simply branch in performWork before the loop.

File: src/isomorphic/classic/class/ReactClass.js [hunk 7]
```diff
@@ -728,6 +728,11 @@ var ReactClassMixin = {
    * type signature and the only use case for this, is to avoid that.
    */
   replaceState: function(newState, callback) {
+    if (this.updater.isFiberUpdater) {
```
Comment (sebmarkbage) [inline src/isomorphic/classic/class/ReactClass.js]: Can you explain why Fiber needs to be special here and take a callback argument instead of a separate `enqueueCallback` call?

If we're going down that route, we should just do the same for the stack reconciler instead.

File: src/renderers/shared/stack/reconciler/__tests__/ReactCompositeComponentState-test.js [hunk 8]
```diff
@@ -148,72 +150,146 @@ describe('ReactCompositeComponent-state', () => {
...
-      ['render', 'blue'],
-      ['componentDidUpdate-currentState', 'blue'],
-      ['componentDidUpdate-prevState', 'blue'],
-      ['forceUpdate', 'blue'],
-      // unmountComponent()
-      // state is available within `componentWillUnmount()`
-      ['componentWillUnmount', 'blue'],
-    ].join('\n'));
+    let expected;
+    if (ReactDOMFeatureFlags.useFiber) {
```
Comment (sebmarkbage) [inline src/renderers/shared/stack/reconciler/__tests__/ReactCompositeComponentState-test.js]: Can we localize this branch to the two places that differ? `...(ReactDOMFeatureFlags.useFiber ? ['foo'] : [])` and `...(!ReactDOMFeatureFlags.useFiber ? ['foo'] : [])`

It is hard to follow what the difference is here.

File: src/isomorphic/classic/class/ReactClass.js (hunk 7 above)
Comment (acdlite) [inline src/isomorphic/classic/class/ReactClass.js]: Took me a while to figure out. If the updater schedules them separately, and the update is synchronous, then the work is already done by the time the callback is scheduled, so the callback doesn't get called until the next flush. This is what I was trying to describe on Friday that is similar to the `componentWillMount` bug we found.

In incremental mode, the flow goes like this:

- (in updater) Schedule update
- (in updater) Schedule callback
- Perform work

In sync mode, it goes like this

- (in updater) Schedule update
- Because the update is synchronous, the work is performed immediately
- (in updater) Schedule callback

I'm actually not sure how Stack deals with this, come to think of it, but this seemed like the most straightforward solution.

Comment (sebmarkbage) [inline src/isomorphic/classic/class/ReactClass.js]: I think the way stack works is that `enqueueCallback` enqueues a second synchronous operation, which doesn't really do anything except calling the callback.

**Comment by acdlite** [comment]
@sebmarkbage Changed `performAndHandleErrors` to branch on the priority level rather than use a first-class function

**Comment by acdlite** [comment]
There are a few cases I'd like to add unit tests for, but don't know how with the current public API. What do you think of a `syncUpdates` API, like the comment here suggests: https://github.com/facebook/react/blob/master/src/renderers/shared/shared/event/ReactGenericBatching.js#L17-L18

**Comment by acdlite** [comment]
Updated the tests to make it clear which parts are different, per @sebmarkbage's feedback.

**Comment by gaearon** [comment]
What is left to do here? Getting rid of `isFiberUpdater`?

**Comment by acdlite** [comment]
I was planning to do that in a separate PR, to unblock other work. So I don't think there's anything left.

File: src/isomorphic/classic/class/ReactClass.js (hunk 7 above)
Comment (sebmarkbage) [inline src/isomorphic/classic/class/ReactClass.js]: Why does this strategy that Stack uses (two separate synchronous reconciliations) work for Fiber? My worry would be that this is covering up another bug.

Comment (acdlite) [inline src/isomorphic/classic/class/ReactClass.js]: There was a bug because `enqueueUpdate` only adding the callback to the queue; it didn't schedule an update on the fiber. This was unobservable previously because the callback was added to the same update queue node as the state update.

I can get it to work by adding that to `enqueueUpdate`, but it felt wrong that you need an extra flush for a callback that is conceptually linked to the update.

Comment (acdlite) [inline src/isomorphic/classic/class/ReactClass.js]: Oops, meant `enqueueCallback`. On my phone so can't edit.

Comment (sebmarkbage) [inline src/isomorphic/classic/class/ReactClass.js]: I split them out in Stack because they look like they are conceptually linked but they're actually not since the callback isn't fired after a particular state update but after all the state updates. It's an API designed without consideration for batching.

At some point I considered making the enqueueCallback API public instead of the argument to setState.

We only have to do an extra flush now because they're synchronous but as soon as we get rid of that it won't cause an extra flush. They'all always be batched.

We shouldn't even need to traverse the parents in schedule update since we've already scheduled this node's priority.

**Comment by acdlite** [comment]
What I worked on today: removing the `isFiberUpdater` hack I was using.

The problem stemmed from the fact that `enqueueSetState` (etc) and `enqueueCallback` were separate calls. In sync mode, this means the state is queued and flushed, then the callback is queued and flushed separately. But previously there was no way to schedule a callback without re-rendering the entire component.

The commit I just pushed reverts the `isFiberUpdater` hack and addresses the issue by

- adding flags to the `updateQueue` that indicate whether the queue has any callbacks or any updates. If there are no updates (only callbacks), we know we can reuse the previous work without re-rendering.
- creating a new effect type for callbacks. This lets us call the callbacks during the commit phase even if there's no update.

**Comment by acdlite** [comment]
Just ran Ben's Fiber test script. There are three regressions, should be minor, looking into them now.

**Comment by acdlite** [comment]
Ooh, I read incorrectly. It's actually only one test that's failing.

**Comment by acdlite** [comment]
Looks like I'm probably not going to figure this out in time to get it merged tonight :( Sorry, @gaearon.

The thing that's blocking me is this assertion: https://github.com/facebook/react/blob/master/src/addons/transitions/__tests__/ReactTransitionGroup-test.js#L101

For some reason, that `setState` callback is being fired before the child is unmounted, so the log equals `['willLeave', 'didLeave']`, without `'willUnmount'` as expected. I'm having a hard time figuring out why, or reproducing in a simpler test case.

File: src/renderers/shared/fiber/ReactFiberScheduler.js [hunk 9]
```diff
@@ -578,22 +649,29 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
...
+      scheduleAnimationWork(root, priorityLevel);
+    } else {
+      scheduleDeferredWork(root, priorityLevel);
       return;
     }
-    scheduleAnimationWork(root, defaultPriority);
   }
 
-  function scheduleUpdate(fiber: Fiber, priorityLevel : PriorityLevel): void {
+  function scheduleUpdate(fiber: Fiber, priorityLevel : ?PriorityLevel): void {
```
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: It looks like we're not really scheduling updates with any priorityLevel so we can just remove this argument completely to simply things.

File: src/renderers/shared/fiber/ReactFiberCommitWork.js [hunk 10]
```diff
@@ -329,41 +328,40 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
...
+              error = tryCallComponentDidUpdate(instance, prevProps, prevState);
+            }
           }
-        } else {
-          if (typeof instance.componentDidUpdate === 'function') {
-            const prevProps = current.memoizedProps;
-            const prevState = current.memoizedState;
-            error = tryCallComponentDidUpdate(instance, prevProps, prevState);
+          if (finishedWork.alternate) {
+            finishedWork.alternate.updateQueue = null;
```
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberCommitWork.js]: Does this need to be reset if there is only a callback but no update? Or is it already `null` in that case?

File: src/renderers/shared/fiber/ReactFiberScheduler.js [hunk 11]
```diff
@@ -506,6 +495,91 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
...
+      lastScheduledRoot = root;
+
+      if (!shouldBatchUpdates) {
+        // Unless in batched mode, perform work immediately
+        performSynchronousWork();
+      }
+    }
+  }
+
+  function performAndHandleErrors(priorityLevel : PriorityLevel, deadline : ?Deadline) {
```
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: Can we make this `null | Deadline`? Not a fan of the last argument accepting `undefined` since the type system won't catch accidental missing args.

File: src/renderers/shared/fiber/ReactFiberScheduler.js [hunk 12]
```diff
@@ -578,22 +649,29 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
     }
   }
 
-  function scheduleWork(root : FiberRoot) {
-    if (defaultPriority === SynchronousPriority) {
-      throw new Error('Not implemented yet');
+  function scheduleWork(root : FiberRoot, priorityLevel : ?PriorityLevel) {
```
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: We should ideally not use `undefined` since this doesn't let us track accidental missing argument. `null | PriorityLevel` would be better.

Ideally we shouldn't have an nullable numbers. This causes them to be boxed (in more optimized compilers than JS VMs).

Maybe there's a way to restructure this into two methods where the other calls this with `priorityContent`?

File: src/renderers/shared/fiber/ReactFiberScheduler.js [hunk 13]
```diff
@@ -619,18 +697,44 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
...
+  function batchedUpdates<A>(fn : () => A) : A {
+    const prev = shouldBatchUpdates;
+    shouldBatchUpdates = true;
+    try {
+      return fn();
+    } finally {
+      shouldBatchUpdates = prev;
+      // If we've exited the batch, perform any scheduled sync work
+      if (!shouldBatchUpdates) {
+        performSynchronousWork();
```
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: Should this be performed if an error is thrown? Currently it would be since it is in `finally`. cc @gaearon

File: src/renderers/shared/fiber/ReactFiberClassComponent.js [hunk 14]
```diff
@@ -27,50 +26,56 @@ var { isMounted } = require('ReactFiberTreeReflection');
...
     }
-    scheduleUpdate(fiber, priorityLevel);
+    scheduleUpdate(fiber);
   }
 
   // Class component state updater
   const updater = {
     isMounted,
-    enqueueSetState(instance, partialState) {
+    enqueueSetState(instance, partialState, callback) {
```
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberClassComponent.js]: These should no longer accept a `callback` argument, right?

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberClassComponent.js]: Yeah, forgot to remove

File: src/renderers/shared/stack/reconciler/__tests__/ReactCompositeComponentState-test.js [hunk 15]
```diff
@@ -213,7 +242,9 @@ describe('ReactCompositeComponent-state', () => {
       // unmountComponent()
       // state is available within `componentWillUnmount()`
       ['componentWillUnmount', 'blue'],
-    ].join('\n'));
+    );
+
+    expect(stateListener.mock.calls.join('\n')).toEqual(expected.join('\n'));
   });
```
Comment (sebmarkbage) [inline src/renderers/shared/stack/reconciler/__tests__/ReactCompositeComponentState-test.js]: It would be nice to have a unit test that covers the case you fixed with the new side-effect type. I.e. `setState({}, cb);` where `shouldComponentUpdate` bails out so there is no `componentDidUpdate`.

**Comment by acdlite** [comment]
Alright I gotta go to bed.

With `git bisect`, I was able the bug down to this line: https://github.com/facebook/react/pull/8127/files#diff-da4b5824156d496f5af2a6372de5901cR78

If I revert that change the TransitionGroup test (unmount child before calling parent callback) passes; but the callback isn't call until/unless there's another next flush, which was the original problem I was having. So something's still not right with these callbacks.

I'll make changes according to review comments and pick up debugging this (hopefully) final bug in the morning.

**Comment by sebmarkbage** [comment]
I think this error actually makes more sense than the previous behavior. The `componentDidUpdate` of ReactTransitionGroup and callback fires at the same time. After that, the unmount that was caused by `componentDidUpdate` commits. I think in stack, callbacks are special cased and happen after any state updates during `componentDidUpdate` have been processed.

We can land this while we figure out what behavior we want.

**Comment by sebmarkbage** [comment]
Accepting pending the comments I mentioned above.

**Comment by gaearon** [comment]
Thanks a lot folks!

File: src/renderers/shared/fiber/ReactFiberScheduler.js (hunk 13 above)
Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: Left a note for myself to check on this in https://github.com/facebook/react/issues/8181.

File: src/renderers/shared/fiber/ReactFiberCommitWork.js (hunk 10 above)
Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberCommitWork.js]: I don't think we need to reset it because we track on the update queue node whether the callback was called already. Are you saying that we should leave it to avoid the extra allocation on the next update?

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberCommitWork.js]: Oh haha oops I didn't read the code correctly. I think it works either way, but the way it is now will avoid an allocation.

**Comment by acdlite** [comment]
Merging once the build passes

File: scripts/fiber/tests-passing.txt [hunk 16]
```diff
@@ -692,6 +691,7 @@ src/renderers/dom/stack/client/__tests__/ReactMount-test.js
 * should warn if mounting into dirty rendered markup
 * should not warn if mounting into non-empty node
 * should warn when mounting into document.body
+* passes the correct callback context
```
Comment (gaearon) [inline scripts/fiber/tests-passing.txt]: Just learned about this a week ago. This is the weirdest part of React API 😄

**Comment by sebmarkbage** [comment]
So, there is a use case for the existing behavior of the callback. It allows you to do work after your parent's `componentDidMount`. Which is what we do for things like controlled components. It'll be interesting to see if this behavior will break something critical.

**Comment by gaearon** [comment]
I fixed up lint. Test probably failed due to my mistake in master which I have fixed there.
